File: frontend-flutter/lib/src/features/tasks/domain/task.dart
import '../../reports/domain/report.dart';

enum TaskStatus {
  newTask('NEW', 'New'),
  assigned('ASSIGNED', 'Assigned'),
  inProgress('IN_PROGRESS', 'In progress'),
  done('DONE', 'Done'),
  pendingReview('PENDING_REVIEW', 'Pending review'),
  approved('APPROVED', 'Approved'),
  closed('CLOSED', 'Closed'),
  cancelled('CANCELLED', 'Cancelled');

  const TaskStatus(this.wireName, this.label);

  final String wireName;
  final String label;

  bool get canAssign =>
      this != TaskStatus.closed && this != TaskStatus.cancelled;

  bool get canClose =>
      this != TaskStatus.closed && this != TaskStatus.cancelled;

  bool get canCancel =>
      this != TaskStatus.closed && this != TaskStatus.cancelled;

  bool get canStart => this == TaskStatus.assigned;

  bool get canComplete => this == TaskStatus.inProgress;

  static TaskStatus fromJson(String value) {
    return TaskStatus.values.firstWhere(
      (status) => status.wireName == value,
      orElse: () => TaskStatus.newTask,
    );
  }
}

class Task {
  const Task({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.status,
    required this.latitude,
    required this.longitude,
    required this.addressText,
    required this.priorityScore,
    required this.assignedStaff,
    required this.createdByOverseer,
    required this.beforePhotoUrl,
    required this.afterPhotoUrl,
    required this.staffNote,
    required this.aiConfidenceScore,
    required this.aiDecision,
    required this.startedAt,
    required this.submittedAt,
    required this.reviewedAt,
    required this.closedAt,
    required this.createdAt,
    required this.updatedAt,
    required this.reportIds,
  });

  final String id;
  final String title;
  final String description;
  final ReportCategory category;
  final TaskStatus status;
  final double latitude;
  final double longitude;
  final String? addressText;
  final int priorityScore;
  final ReportUserSummary? assignedStaff;
  final ReportUserSummary? createdByOverseer;
  final String? beforePhotoUrl;
  final String? afterPhotoUrl;
  final String? staffNote;
  final double? aiConfidenceScore;
  final String? aiDecision;
  final DateTime? startedAt;
  final DateTime? submittedAt;
  final DateTime? reviewedAt;
  final DateTime? closedAt;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<String> reportIds;

  String get locationLabel {
    final address = addressText?.trim();
    if (address != null && address.isNotEmpty) {
      return address;
    }
    return '${latitude.toStringAsFixed(4)}, ${longitude.toStringAsFixed(4)}';
  }

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String,
      category: ReportCategory.fromJson(json['category'] as String),
      status: TaskStatus.fromJson(json['status'] as String),
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      addressText: json['addressText'] as String?,
      priorityScore: json['priorityScore'] as int? ?? 0,
      assignedStaff: _userSummary(json['assignedStaff']),
      createdByOverseer: _userSummary(json['createdByOverseer']),
      beforePhotoUrl: json['beforePhotoUrl'] as String?,
      afterPhotoUrl: json['afterPhotoUrl'] as String?,
      staffNote: json['staffNote'] as String?,
      aiConfidenceScore: (json['aiConfidenceScore'] as num?)?.toDouble(),
      aiDecision: json['aiDecision'] as String?,
      startedAt: _dateTime(json['startedAt']),
      submittedAt: _dateTime(json['submittedAt']),
      reviewedAt: _dateTime(json['reviewedAt']),
      closedAt: _dateTime(json['closedAt']),
      createdAt: _dateTime(json['createdAt']) ?? DateTime.now(),
      updatedAt: _dateTime(json['updatedAt']) ?? DateTime.now(),
      reportIds: (json['reportIds'] as List<dynamic>? ?? const <dynamic>[])
          .map((id) => id as String)
          .toList(growable: false),
    );
  }

  Task copyWith({
    String? id,
    String? title,
    String? description,
    ReportCategory? category,
    TaskStatus? status,
    double? latitude,
    double? longitude,
    String? addressText,
    int? priorityScore,
    ReportUserSummary? assignedStaff,
    ReportUserSummary? createdByOverseer,
    String? beforePhotoUrl,
    String? afterPhotoUrl,
    String? staffNote,
    double? aiConfidenceScore,
    String? aiDecision,
    DateTime? startedAt,
    DateTime? submittedAt,
    DateTime? reviewedAt,
    DateTime? closedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
    List<String>? reportIds,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      category: category ?? this.category,
      status: status ?? this.status,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      addressText: addressText ?? this.addressText,
      priorityScore: priorityScore ?? this.priorityScore,
      assignedStaff: assignedStaff ?? this.assignedStaff,
      createdByOverseer: createdByOverseer ?? this.createdByOverseer,
      beforePhotoUrl: beforePhotoUrl ?? this.beforePhotoUrl,
      afterPhotoUrl: afterPhotoUrl ?? this.afterPhotoUrl,
      staffNote: staffNote ?? this.staffNote,
      aiConfidenceScore: aiConfidenceScore ?? this.aiConfidenceScore,
      aiDecision: aiDecision ?? this.aiDecision,
      startedAt: startedAt ?? this.startedAt,
      submittedAt: submittedAt ?? this.submittedAt,
      reviewedAt: reviewedAt ?? this.reviewedAt,
      closedAt: closedAt ?? this.closedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      reportIds: reportIds ?? this.reportIds,
    );
  }
}

class TaskDraft {
  const TaskDraft({
    required this.title,
    required this.description,
    required this.category,
    required this.latitude,
    required this.longitude,
    required this.addressText,
    required this.priorityScore,
    required this.assignedStaffId,
    required this.beforePhotoUrl,
    required this.afterPhotoUrl,
    required this.staffNote,
    required this.reportIds,
  });

  final String title;
  final String description;
  final ReportCategory category;
  final double latitude;
  final double longitude;
  final String? addressText;
  final int priorityScore;
  final String? assignedStaffId;
  final String? beforePhotoUrl;
  final String? afterPhotoUrl;
  final String? staffNote;
  final List<String> reportIds;

  Map<String, Object?> toCreateJson() {
    return <String, Object?>{
      'title': title,
      'description': description,
      'category': category.wireName,
      'latitude': latitude,
      'longitude': longitude,
      'addressText': addressText,
      'priorityScore': priorityScore,
      'assignedStaffId': assignedStaffId,
      'beforePhotoUrl': beforePhotoUrl,
      'reportIds': reportIds,
    };
  }

  Map<String, Object?> toUpdateJson() {
    return <String, Object?>{
      'title': title,
      'description': description,
      'category': category.wireName,
      'latitude': latitude,
      'longitude': longitude,
      'addressText': addressText,
      'priorityScore': priorityScore,
      'beforePhotoUrl': beforePhotoUrl,
      'afterPhotoUrl': afterPhotoUrl,
      'staffNote': staffNote,
      'reportIds': reportIds,
    };
  }
}

class TaskCompletionDraft {
  const TaskCompletionDraft({
    required this.afterPhotoUrl,
    required this.staffNote,
  });

  final String? afterPhotoUrl;
  final String? staffNote;

  Map<String, Object?> toJson() {
    return <String, Object?>{
      'afterPhotoUrl': afterPhotoUrl,
      'staffNote': staffNote,
    };
  }
}

ReportUserSummary? _userSummary(Object? value) {
  if (value is Map<String, dynamic>) {
    return ReportUserSummary.fromJson(value);
  }
  return null;
}

DateTime? _dateTime(Object? value) {
  if (value is String && value.isNotEmpty) {
    return DateTime.parse(value);
  }
  return null;
}

File: backend-spring/src/main/java/com/smartcity/reports/task/domain/Task.java
package com.smartcity.reports.task.domain;

import com.smartcity.reports.issue.IssueCategory;
import com.smartcity.reports.report.domain.Report;
import com.smartcity.reports.user.domain.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import org.locationtech.jts.geom.Point;

import java.time.Instant;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "tasks")
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, length = 120)
    private String title;

    @Column(nullable = false, columnDefinition = "text")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 40)
    private IssueCategory category;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private TaskStatus status;

    @Column(nullable = false)
    private double latitude;

    @Column(nullable = false)
    private double longitude;

    @Column(name = "address_text", length = 255)
    private String addressText;

    @Column(name = "priority_score", nullable = false)
    private int priorityScore;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_staff_id")
    private User assignedStaff;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_overseer_id")
    private User createdByOverseer;

    @Column(name = "before_photo_url", length = 2048)
    private String beforePhotoUrl;

    @Column(name = "after_photo_url", length = 2048)
    private String afterPhotoUrl;

    @Column(name = "staff_note", columnDefinition = "text")
    private String staffNote;

    @Column(name = "ai_confidence_score")
    private Double aiConfidenceScore;

    @Column(name = "ai_decision", length = 120)
    private String aiDecision;

    @Column(name = "started_at")
    private Instant startedAt;

    @Column(name = "submitted_at")
    private Instant submittedAt;

    @Column(name = "reviewed_at")
    private Instant reviewedAt;

    @Column(name = "closed_at")
    private Instant closedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @Column(
            name = "location",
            insertable = false,
            updatable = false,
            columnDefinition = "geometry(Point,4326)"
    )
    private Point location;

    @ManyToMany
    @JoinTable(
            name = "task_reports",
            joinColumns = @JoinColumn(name = "task_id"),
            inverseJoinColumns = @JoinColumn(name = "report_id")
    )
    private Set<Report> reports = new LinkedHashSet<>();

    protected Task() {
    }

    public Task(
            String title,
            String description,
            IssueCategory category,
            double latitude,
            double longitude,
            String addressText,
            int priorityScore,
            User assignedStaff,
            User createdByOverseer,
            String beforePhotoUrl
    ) {
        this.title = title;
        this.description = description;
        this.category = category;
        this.latitude = latitude;
        this.longitude = longitude;
        this.addressText = addressText;
        this.priorityScore = Math.max(0, priorityScore);
        this.assignedStaff = assignedStaff;
        this.createdByOverseer = createdByOverseer;
        this.beforePhotoUrl = beforePhotoUrl;
        this.status = assignedStaff == null ? TaskStatus.NEW : TaskStatus.ASSIGNED;
    }

    @PrePersist
    void prePersist() {
        Instant now = Instant.now();
        if (createdAt == null) {
            createdAt = now;
        }
        if (updatedAt == null) {
            updatedAt = now;
        }
        if (status == null) {
            status = assignedStaff == null ? TaskStatus.NEW : TaskStatus.ASSIGNED;
        }
    }

    @PreUpdate
    void preUpdate() {
        updatedAt = Instant.now();
    }

    public void updateDetails(
            String title,
            String description,
            IssueCategory category,
            double latitude,
            double longitude,
            String addressText,
            int priorityScore,
            String beforePhotoUrl,
            String afterPhotoUrl,
            String staffNote
    ) {
        this.title = title;
        this.description = description;
        this.category = category;
        this.latitude = latitude;
        this.longitude = longitude;
        this.addressText = addressText;
        this.priorityScore = Math.max(0, priorityScore);
        this.beforePhotoUrl = beforePhotoUrl;
        this.afterPhotoUrl = afterPhotoUrl;
        this.staffNote = staffNote;
    }

    public void assign(User staff) {
        assignedStaff = staff;
        status = TaskStatus.ASSIGNED;
    }

    public void start(Instant now) {
        status = TaskStatus.IN_PROGRESS;
        startedAt = now;
    }

    public void complete(Instant now, String afterPhotoUrl, String staffNote) {
        status = TaskStatus.DONE;
        submittedAt = now;
        this.afterPhotoUrl = afterPhotoUrl;
        if (staffNote != null) {
            this.staffNote = staffNote;
        }
        aiConfidenceScore = null;
        aiDecision = null;
    }

    public void approve(Instant now) {
        status = TaskStatus.APPROVED;
        reviewedAt = now;
    }

    public void close(Instant now) {
        status = TaskStatus.CLOSED;
        closedAt = now;
    }

    public void cancel() {
        status = TaskStatus.CANCELLED;
    }

    public void linkReport(Report report) {
        reports.add(report);
    }

    public void unlinkReport(Report report) {
        reports.remove(report);
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public IssueCategory getCategory() {
        return category;
    }

    public TaskStatus getStatus() {
        return status;
    }

    public double getLatitude() {
        return latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public String getAddressText() {
        return addressText;
    }

    public int getPriorityScore() {
        return priorityScore;
    }

    public User getAssignedStaff() {
        return assignedStaff;
    }

    public User getCreatedByOverseer() {
        return createdByOverseer;
    }

    public String getBeforePhotoUrl() {
        return beforePhotoUrl;
    }

    public String getAfterPhotoUrl() {
        return afterPhotoUrl;
    }

    public String getStaffNote() {
        return staffNote;
    }

    public Double getAiConfidenceScore() {
        return aiConfidenceScore;
    }

    public String getAiDecision() {
        return aiDecision;
    }

    public Instant getStartedAt() {
        return startedAt;
    }

    public Instant getSubmittedAt() {
        return submittedAt;
    }

    public Instant getReviewedAt() {
        return reviewedAt;
    }

    public Instant getClosedAt() {
        return closedAt;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Instant updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Point getLocation() {
        return location;
    }

    public Set<Report> getReports() {
        return reports;
    }
}
